### DISCUSSION TITLE
Ai copilot assistant

### DISCUSSION THREAD

**Discussion by friuns2** [phcode-dev/phoenix#1192] [Ideas]
Add Ai copilot assistant 
Would work like in vercel https://v0.dev/ with live preview

**Comment by abose** [comment]
Working on this. Most likeley avilable by first half of 2024.

**Reply by ijomeli** [reply]
could we maybe have a code editor that doesn't send all my stuff to openAI actually, that would be nice

**Reply by abose** [reply]
Scheduled for development only in the later half of 2024. We have been experimenting with llama and other models, but havent concluded which one we will be finally using.

**Reply by AmberShards** [reply]
There's not even the ability to hide the intrusive AI tab unless you turn off the file sidebar completely. I've posted an idea to resolve this.

**Comment by pashas84** [comment]
any update on this?

**Reply by devvaannsh** [reply]
Hey @pashas84, we're actively working on it. Our basic prototype is ready too, but this will take some time. It should be available for use by the May/June release if things go as planned.

**Comment by hankbinkle** [comment]
My experience with AI has been pretty positive but work is stymied by endless cut and paste.  Often a round of changes involves many steps to verify the code.  This is particularly tedious when the AI code fails due to something like a simple syntax error or null value.  I would love it if a Phoenix AI tool could auto iterate certain responses.  For example, when a portion of AI code generates a console error, AI simply addresses the error immediately and continues to do so until no errors are returned.

Thanks for all your amazing work!

**Reply by devvaannsh** [reply]
Hey @hankbinkle , the AI work is currently in progress. We'll ship the AI assistant in the next update which will most probably come by the mid of next month. 

Btw, thanks for the suggestion. We'll definitely look into it and see if it can be done. Also, are there any existing AI tools that does what you mentioned? That'll give us a better idea of the workflow.

**Comment by hankbinkle** [comment]
Oh that is great! No, I just use Gemini or ChatGpt and copy/paste back and
forth! Funny that with such power at our fingertips, cut and paste is still
a thing!

Thanks!

Thanks,

Rob

Rob Zinn
c:904.377.9854
***@***.***

Sent via Superhuman iOS ***@***.***>

**Reply by devvaannsh** [reply]
Interestingly, Phoenix AI will solve that problem. Phoenix AI will automatically edit files, add new files and whatever change needed as per your prompt. So that will completely eliminate the copy/paste chores. 🚀

**Comment by TroyRobinson** [comment]
Thank you for your hard work on Phoenix AI! I believe the combination of LLMs + Phoenix' bi-directional element selection could be very powerful for web app prototyping and iteration. If you want a UX/UI professional to help with the early versions, I'm happy to contribute. @abose @devvaannsh 

(I built an LLM coding extension for Monaco/VSCode based editors and was about to expand it to PHCode until I discovered you're already working on the same end!)

**Comment by TroyRobinson** [comment]
Question: will the AI features be accessible via the extension API (so we can expand on the features)?

**Reply by abose** [reply]
The core AI framework is being built, but we thought of adding extensibility after the main AI is out in the next release for simplicity. 

For extensibility, can you give an idea of what all features or extensible parts you would like to have. A few examples will help me understand what extensions need access to.

**Reply by TroyRobinson** [reply]
@abose Thank you for the clarification. For example, say an AI extension user wants to (video below):
1. click a container element in the rendered preview
2. automatically select all the related code in the editor
3. maybe use a light LLM to quickly, additionally select related code in the file.
4. type a quick design modification prompt in an inline prompt popup (e.g. like control+k in Cursor)
5. send off that code to the LLM and receive updated code.
6. replace selections in the editor with the updated code.

I think many of these abilities are already possible via your current API although I imagine much of the new LLM usage and prompt interfaces could be made a seamlessly integrated part of the editor such that others don't need to rebuild the features to do something like the above. Does that make sense?

Video example walkthrough (~30sec):
https://github.com/user-attachments/assets/b5d8ff80-688b-45ec-992c-286cc915424d
- [High-resolution video](https://videy.co/v/?id=juC0NPyr1)

**Reply by abose** [reply]
Thanks for the video and details. Just curious on what is the editor shown in the video?

**Reply by TroyRobinson** [reply]
@abose You're welcome! The editor I was using is [Utopia (Github)](https://github.com/concrete-utopia/utopia), a really ahead-of-its-time and very fast bi-directional editor & designer (like Figma & VSCode merged). React-as-the-source-of-truth designing.
+ Plus, a prototype Chrome extension I've been working on to integrate LLM design edits.

**Comment by hankbinkle** [comment]
Hey Troy. If you have UI/UX experience, I might have a project for you if
you are interested?

Thanks,

Rob

Rob Zinn
c:904.377.9854
***@***.***

Sent via Superhuman iOS ***@***.***>

**Comment by afalina7** [comment]
было бы интересно попробовать
